var/spack/repos/builtin/packages/py-netcdf4/package.py

```diff
@@ -36,15 +38,15 @@ class PyNetcdf4(PythonPackage):
     depends_on("py-setuptools-scm@3.4:+toml", when="@1.7:", type="build")
     depends_on("py-cftime", type=("build", "run"))
     depends_on("py-certifi", when="@1.6.5:", type=("build", "run"))
-    depends_on("py-numpy", when="@1.6.5:", type=("build", "link", "run"))
+    depends_on("py-numpy", type=("build", "link", "run"))
+    depends_on("py-numpy@2.0:", when="@1.7.1:", type=("build", "link", "run"))
     depends_on("py-numpy@1.9:", when="@1.5.4:1.6.2", type=("build", "link", "run"))
-    depends_on("py-numpy@1.7:", type=("build", "link", "run"))
     # https://github.com/Unidata/netcdf4-python/pull/1317
     depends_on("py-numpy@:1", when="@:1.6", type=("build", "link", "run"))
     depends_on("py-mpi4py", when="+mpi", type=("build", "run"))
-    depends_on("netcdf-c", when="-mpi")
+    depends_on("netcdf-c~mpi", when="~mpi")  # netcdf-c has +mpi by default
```

Comment:
Several years ago I was told that Spack was a package manager for HPC. Therefore, all packages that had the `mpi` variant `default`ed it to `True`. If that is still the case, the comment is redundant (same for `hdf5` below).

I would also like to make it clear that requiring `netcdf-c~mpi` and `hdf5~mpi` when `~mpi` is not a convenience feature that makes it easier for the users to get rid of the `mpi` dependency. We have the variant propagation for that (see [here](https://spack.readthedocs.io/en/latest/basic_usage.html#specs-dependencies)), e.g. `spack install py-netcdf4 ~~mpi`. The reason for having those constraints is questionable implementation details of the `py-netcdf4`'s building script.

First, the non-MPI functionality of `netcdf-c` is available even if it is built with `+mpi`. However, the fact that `netcdf-c` is built with `+mpi` is a trigger for [`setup.py`](https://github.com/Unidata/netcdf4-python/blob/9426a684d3bd7b6c0c9b4d3c3f8531f30583d1e2/setup.py#L58C45-L58C60) to start building with MPI support (it wants to have `py-mpi4py` and so on). Therefore, the only option to get `py-netcdf4~mpi` is to build it against `netcdf-c~mpi`.

Second, `py-netcdf4` should not care about the backend of the backend, i.e. `hdf5`. The `netcdf-c` package itself can be built with `~mpi` even if `^hdf5+mpi`. This is not the case for `py-netcdf4`. For whatever reason, it wants to know the version of `hdf5` behind `netcdf-c` and includes `H5public.h` for that (see [here](https://github.com/Unidata/netcdf4-python/blob/9426a684d3bd7b6c0c9b4d3c3f8531f30583d1e2/include/netCDF4.pxi#L7-L9)). The problem is that `H5public.h` includes `mpi.h`, so it has to be available and, technically, `mpi` should be a dependency of every package that includes the header and is built against `hdf5+mpi`. I don't know what happens in the edge cases when `mpi` and `hdf5` are `externals`, but if everything is built with Spack, the compiler wrapper injects the required `-I` flag pointing to `mpi.h` and `#include <H5public.h>` works even without the MPI compiler wrapper and having `mpi` as a direct dependency. Alas, there is [this](https://github.com/Unidata/netcdf4-python/blob/9426a684d3bd7b6c0c9b4d3c3f8531f30583d1e2/include/no_parallel_support_imports.pxi.in#L3-L10), which conflicts with definitions from `mpi.h`. Therefore, the only option to get `py-netcdf4~mpi` is to build it against `hdf5~mpi`.

Reply:
>Therefore, all packages that had the mpi variant defaulted it to True

Ah this is good to know. Makes sense. I've removed the comments in light of this.

Thanks for the additional details. I had surmised this was what was going on, but your confirmation of a couple details is super helpful. To me it seems like we are constrained by how `py-netcdf4` works, and there isn't much else we can do. If so, is there any outstanding change? Is it worth linking to your comment in the package as a comment?